service/worker/scanner/shardscanner/activities.go

```diff
@@ -0,0 +1,336 @@
+			ActivityBatchSize:       dc.ActivityBatchSize(),
+		},
+	}
+
+	if ctx.Hooks != nil && ctx.Hooks.Config != nil {
+		result.CustomScannerConfig = ctx.Hooks.Config(ctx)
+	}
+
+	overwrites := params.Overwrites.GenericScannerConfig
+	if overwrites.Enabled != nil {
+		result.GenericScannerConfig.Enabled = *overwrites.Enabled
+	}
+	if overwrites.Concurrency != nil {
+		result.GenericScannerConfig.Concurrency = *overwrites.Concurrency
+	}
+	if overwrites.PageSize != nil {
+		result.GenericScannerConfig.PageSize = *overwrites.PageSize
+	}
+	if overwrites.BlobstoreFlushThreshold != nil {
+		result.GenericScannerConfig.BlobstoreFlushThreshold = *overwrites.BlobstoreFlushThreshold
+	}
+
+	if overwrites.ActivityBatchSize != nil {
+		result.GenericScannerConfig.ActivityBatchSize = *overwrites.ActivityBatchSize
+	}
+
+	if params.Overwrites.CustomScannerConfig != nil {
+		result.CustomScannerConfig = *params.Overwrites.CustomScannerConfig
+	}
+
+	return result, nil
+}
+
+// ScanShardActivity will scan a collection of shards for invariant violations.
+func ScanShardActivity(
+	activityCtx context.Context,
+	params ScanShardActivityParams,
+) ([]ScanReport, error) {
+	heartbeatDetails := ScanShardHeartbeatDetails{
+		LastShardIndexHandled: -1,
+		Reports:               nil,
+	}
+	if activity.HasHeartbeatDetails(activityCtx) {
+		if err := activity.GetHeartbeatDetails(activityCtx, &heartbeatDetails); err != nil {
+			return nil, err
+		}
+	}
+	for i := heartbeatDetails.LastShardIndexHandled + 1; i < len(params.Shards); i++ {
+		currentShardID := params.Shards[i]
+		shardReport, err := scanShard(activityCtx, params, currentShardID, heartbeatDetails)
+		if err != nil {
+			return nil, err
+		}
+		heartbeatDetails = ScanShardHeartbeatDetails{
+			LastShardIndexHandled: i,
+			Reports:               append(heartbeatDetails.Reports, *shardReport),
+		}
+	}
+	return heartbeatDetails.Reports, nil
+}
+
+func scanShard(
+	activityCtx context.Context,
+	params ScanShardActivityParams,
+	shardID int,
+	heartbeatDetails ScanShardHeartbeatDetails,
+) (*ScanReport, error) {
+	ctx := activityCtx.Value(params.ContextKey).(Context)
+	scope := ctx.Scope.Tagged(metrics.ActivityTypeTag(params.ContextKey.String()))
+	sw := scope.StartTimer(metrics.CadenceLatency)
+	defer sw.Stop()
+
+	if ctx.Hooks == nil {
+		return nil, errors.New("scanner hooks are not provided")
+	}
+
+	resources := ctx.Resource
+	execManager, err := resources.GetExecutionManager(shardID)
+	if err != nil {
+		scope.IncCounter(metrics.CadenceFailures)
+		return nil, err
+	}
+
+	pr := persistence.NewPersistenceRetryer(execManager, resources.GetHistoryManager(), c.CreatePersistenceRetryPolicy())
+
+	scanner := NewScanner(
+		shardID,
+		ctx.Hooks.Iterator(pr, params, *ctx.Config),
+		resources.GetBlobstoreClient(),
+		params.BlobstoreFlushThreshold,
+		ctx.Hooks.Manager(pr, params, *ctx.Config),
+		func() { activity.RecordHeartbeat(activityCtx, heartbeatDetails) },
+	)
+	report := scanner.Scan(activityCtx)
```

Comment:
It seems to me that report should be a struct composed of two types: GenericReportInfo and ScanSpecificyInfo

Just like you have for config.

Reply:
There are parts of the report that deal only at the level of if shard was successful or not but there are also going to be parts of the report that are specific to the type of scan being run. We can do this in a follow up diff because this one is already very large.